src/wayland/data_device/mod.rs

```diff
@@ -29,285 +27,180 @@
 
 /// Events that are generated by interactions of the clients with the data device
-#[derive(Debug)]
-pub enum DataDeviceEvent {
+pub trait DataDeviceHandler: Sized {
+    /// [DataDeviceState] getter
+    fn data_device_state(&self) -> &DataDeviceState;
+
+    /// Action chooser for DnD negociation
+    fn action_choice(&mut self, available: DndAction, preferred: DndAction) -> DndAction {
+        default_action_chooser(available, preferred)
+    }
+
     /// A client has set the selection
-    NewSelection(Option<wl_data_source::WlDataSource>),
+    #[allow(unused_variables)]
+    fn new_selection(&mut self, source: Option<WlDataSource>) {}
+
+    /// A client requested to read the server-set selection
+    ///
+    /// * `mime_type` - the requested mime type
+    /// * `fd` - the fd to write into
+    #[allow(unused_variables)]
+    fn send_selection(&mut self, mime_type: String, fd: RawFd) {}
+}
+
+/// Events that are generated during client initiated drag'n'drop
+pub trait ClientDndGrabHandler: Sized {
     /// A client started a drag'n'drop as response to a user pointer action
-    DnDStarted {
-        /// The data source provided by the client
-        ///
-        /// If it is `None`, this means the DnD is restricted to surfaces of the
-        /// same client and the client will manage data transfer by itself.
-        source: Option<wl_data_source::WlDataSource>,
-        /// The icon the client requested to be used to be associated with the cursor icon
-        /// during the drag'n'drop.
-        icon: Option<wl_surface::WlSurface>,
-        /// The seat on which the DnD operation was started
-        seat: Seat,
-    },
+    ///
+    /// * `source` - The data source provided by the client.
+    ///              If it is `None`, this means the DnD is restricted to surfaces of the
+    ///              same client and the client will manage data transfer by itself.
+    /// * `icon` - The icon the client requested to be used to be associated with the cursor icon
+    ///            during the drag'n'drop.
+    /// * `seat` - The seat on which the DnD operation was started
+    #[allow(unused_variables)]
+    fn started(&mut self, source: Option<WlDataSource>, icon: Option<WlSurface>, seat: Seat<Self>) {}
+
     /// The drag'n'drop action was finished by the user releasing the buttons
     ///
     /// At this point, any pointer icon should be removed.
     ///
     /// Note that this event will only be generated for client-initiated drag'n'drop session.
-    DnDDropped {
-        /// The seat on which the DnD action was finished.
-        seat: Seat,
-    },
-    /// A client requested to read the server-set selection
-    SendSelection {
-        /// the requested mime type
-        mime_type: String,
-        /// the fd to write into
-        fd: RawFd,
-    },
+    ///
+    /// * `seat` - The seat on which the DnD action was finished.
+    #[allow(unused_variables)]
+    fn dropped(&mut self, seat: Seat<Self>) {}
```

Comment:
Warning can be suppressed by naming the variable `_seat`

Reply:
The idea was to not call it with _ so autocompleation works better for those, but yeah maybe it's not worth it.

Reply:
In those cases you can also just do `let _ = seat;` inside the implementation.